### ISSUE TITLE
Regression in 32bit build

### ISSUE THREAD

**Issue by mmikulicic** [cesanta/v7#335]
Regression introduced in 738b340:

```
$ make -C tests test_m32 CFLAGS_EXTRA="-DV7_ENABLE_COMPACTING_GC" TEST_FILTER=ecma
CC      unit_test_m32
RUN     unit_test_m32 ecma
make: *** [test_m32] Segmentation fault: 11
```

**Comment by mmikulicic** [comment]
```
rm v7 -f && make asan_v7 CFLAGS_EXTRA="-m32 -DV7_ENABLE_COMPACTING_GC" V7_ARGS="tests/ecma_driver.js test.js"
```

```
=================================================================
==98086==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x01d011f7 at pc 0x000e392b bp 0xbff3d898 sp 0xbff3d894
READ of size 8 at 0x01d011f7 thread T0
    #0 0xe392a in a_partition /Users/mkm/Projects/cesanta/v7/src/array.c:97:3
    #1 0xe309b in a_qsort /Users/mkm/Projects/cesanta/v7/src/array.c:112:13
    #2 0xe22cd in a_sort /Users/mkm/Projects/cesanta/v7/src/array.c:136:5
    #3 0xd5665 in Array_sort /Users/mkm/Projects/cesanta/v7/src/array.c:153:10
    #4 0x19878c in i_eval_call /Users/mkm/Projects/cesanta/v7/src/interpreter.c:958:11
    #5 0x18cb19 in i_eval_expr /Users/mkm/Projects/cesanta/v7/src/interpreter.c:650:13
    #6 0x17edfe in i_eval_expr /Users/mkm/Projects/cesanta/v7/src/interpreter.c:292:12
    #7 0x16d45e in i_eval_stmt /Users/mkm/Projects/cesanta/v7/src/interpreter.c:1075:26
    #8 0x16963f in i_eval_stmts /Users/mkm/Projects/cesanta/v7/src/interpreter.c:1041:11
    #9 0x16cd23 in i_eval_stmt /Users/mkm/Projects/cesanta/v7/src/interpreter.c:1068:13
    #10 0x16ad33 in v7_exec_with /Users/mkm/Projects/cesanta/v7/src/interpreter.c:1541:7
    #11 0x17ac66 in v7_exec /Users/mkm/Projects/cesanta/v7/src/interpreter.c:1555:10
    #12 0x17bf9b in v7_exec_file /Users/mkm/Projects/cesanta/v7/src/interpreter.c:1582:11
    #13 0x1d32cc in main /Users/mkm/Projects/cesanta/v7/src/main.c:100:16
    #14 0x934d0700 in start (/usr/lib/system/libdyld.dylib+0x3700)
    #15 0x2 (<unknown module>)

0x01d011f7 is located 6 bytes to the right of 1-byte region [0x01d011f0,0x01d011f1)
allocated by thread T0 here:
    #0 0x26bd9a in wrap_malloc (/usr/local/lib/llvm-3.5/lib/clang/3.5.0/lib/darwin/libclang_rt.asan_osx_dynamic.dylib+0x2fd9a)
    #1 0xe16b8 in a_sort /Users/mkm/Projects/cesanta/v7/src/array.c:121:26
    #2 0xd5665 in Array_sort /Users/mkm/Projects/cesanta/v7/src/array.c:153:10
    #3 0x19878c in i_eval_call /Users/mkm/Projects/cesanta/v7/src/interpreter.c:958:11
    #4 0x18cb19 in i_eval_expr /Users/mkm/Projects/cesanta/v7/src/interpreter.c:650:13
    #5 0x17edfe in i_eval_expr /Users/mkm/Projects/cesanta/v7/src/interpreter.c:292:12
    #6 0x16d45e in i_eval_stmt /Users/mkm/Projects/cesanta/v7/src/interpreter.c:1075:26
    #7 0x16963f in i_eval_stmts /Users/mkm/Projects/cesanta/v7/src/interpreter.c:1041:11
    #8 0x16cd23 in i_eval_stmt /Users/mkm/Projects/cesanta/v7/src/interpreter.c:1068:13
    #9 0x16ad33 in v7_exec_with /Users/mkm/Projects/cesanta/v7/src/interpreter.c:1541:7
    #10 0x17ac66 in v7_exec /Users/mkm/Projects/cesanta/v7/src/interpreter.c:1555:10
    #11 0x17bf9b in v7_exec_file /Users/mkm/Projects/cesanta/v7/src/interpreter.c:1582:11
    #12 0x1d32cc in main /Users/mkm/Projects/cesanta/v7/src/main.c:100:16
    #13 0x934d0700 in start (/usr/lib/system/libdyld.dylib+0x3700)
    #14 0x2 (<unknown module>)

SUMMARY: AddressSanitizer: heap-buffer-overflow /Users/mkm/Projects/cesanta/v7/src/array.c:97 a_partition
Shadow bytes around the buggy address:
  0x203a01e0: fa fa 00 04 fa fa 00 04 fa fa fd fd fa fa fd fa
  0x203a01f0: fa fa fd fa fa fa 00 04 fa fa 00 04 fa fa 00 04
  0x203a0200: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x203a0210: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x203a0220: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
=>0x203a0230: fa fa fa fa fa fa fa fa fa fa fa fa fa fa[01]fa
  0x203a0240: fa fa 00 04 fa fa 00 04 fa fa 00 04 fa fa fd fd
  0x203a0250: fa fa fd fa fa fa fd fa fa fa 00 04 fa fa 00 04
  0x203a0260: fa fa 00 04 fa fa 00 04 fa fa 00 04 fa fa 00 04
  0x203a0270: fa fa 00 04 fa fa 00 04 fa fa 00 04 fa fa 00 04
  0x203a0280: fa fa 00 04 fa fa 00 04 fa fa 00 04 fa fa 00 04
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Heap right redzone:      fb
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack partial redzone:   f4
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  ASan internal:           fe
==98086==ABORTING
```

Where `test.js` is:

```
/**
 * Check ToUint32(length) for non Array objects
 *
 * @path ch15/15.4/15.4.4/15.4.4.11/S15.4.4.11_A4_T1.js
 * @description length = 4294967296
 */

var obj = {};
obj.sort = Array.prototype.sort;
obj[0] = "x";
obj[4294967295] = "y";
obj.length = 4294967296;

//CHECK#1
if (obj.sort() !== obj) {
  $ERROR('#1: var obj = {}; obj.sort = Array.prototype.sort; obj[0] = "x"; obj[4294967295] = "y"; obj.length = 4294967296; obj.sort() === obj. Actual: ' + (obj.sort()));
}

//CHECK#2
if (obj.length !== 4294967296) {
  $ERROR('#2: var obj = {}; obj.sort = Array.prototype.sort; obj[0] = "x"; obj[4294967295] = "y"; obj.length = 4294967296; obj.sort(); obj.length === 4294967296. Actual: ' + (obj.length));
}

//CHECK#3
if (obj[0] !== "x") {
  $ERROR('#3: var obj = {}; obj.sort = Array.prototype.sort; obj[0] = "x"; obj[4294967295] = "y"; obj.length = 4294967296; obj.sort(); obj[0] == "x"');
}

//CHECK#4
if (obj[4294967295] !== "y") {
  $ERROR('#4: var obj = {}; obj.sort = Array.prototype.sort; obj[] = "x"; obj[4294967295] = "y"; obj.length = 4294967296; obj.sort(); obj[4294967295] == "y"');
}
```

**Comment by mmikulicic** [comment]
Caused by signed integer overflow in array length.

**Comment by mmikulicic** [comment]
Fixed in 575035712494
